### PR TITLE
Notebooks

### PR DESCRIPTION
**Pull request by dizcology**
With some small updates to the census/estimator sample, its task.py can be automatically converted to a notebook that is import/runnable on Colab as well as local.

This PR includes the prototype tool that handles .py to .ipynb conversion.

---
This change is [Reviewable](https://reviewable.io/reviews/googlecloudplatform/cloudml-samples/277)

### CODE REVIEW

File: census/estimator/trainer/task.py [hunk 1]
```diff
@@ -82,7 +82,7 @@ def run_experiment(hparams):
       '--train-files',
       help='GCS or local paths to training data',
       nargs='+',
-      required=True
+      default=['gs://cloud-samples-data/ml-engine/census/data/adult.data.csv']
```
Comment (gogasca) [inline census/estimator/trainer/task.py]: I submitted a new PR https://github.com/GoogleCloudPlatform/cloudml-samples/pull/276/commits/0a43f62e6ff3f805e25906f62698aa47c5a87347 which does't require the list parameter. I will approve for now.
If you can approve later that PR and then update the changes here will be good too.

File: census/estimator/trainer/task.py [hunk 2]
```diff
@@ -110,7 +110,7 @@ def run_experiment(hparams):
       '--eval-files',
       help='GCS or local paths to evaluation data',
       nargs='+',
-      required=True
+      default=['gs://cloud-samples-data/ml-engine/census/data/adult.test.csv']
```
Comment (gogasca) [inline census/estimator/trainer/task.py]: Same as above

File: census/estimator/trainer/task.ipynb [hunk 3]
```diff
@@ -0,0 +1,242 @@
+{
```
Comment (gogasca) [inline census/estimator/trainer/task.ipynb]: If we can add some Markup text/Documentation in terms what the Notebook is doing will be great

File: tools/to_ipynb.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
+import os
```
Comment (gogasca) [inline tools/to_ipynb.py]: Reorder imports and simplify multiple imports:

import os
import re
import nbformat
from nbformat.v4 import new_code_cell
from nbformat.v4 import new_markdown_cell
from nbformat.v4 import new_notebook
from redbaron import RedBaron
import yaml

Comment (gogasca) [inline tools/to_ipynb.py]: I don't want to enforce any practice or similar, but I normally use Cider which formats code quite nicely Using PEP-8 standard.
I add this file to this cl/216603659

File: tools/to_ipynb.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+    if prev_type == cur_type and cur_type in concat_types:
+        result = True
+
+    if prev_type in import_types and cur_type in import_types:
+        result = True
+
+    return result
+
+def py_to_ipynb(root, path, py_filename, git_clone=False, remove={}):
```
Comment (gogasca) [inline tools/to_ipynb.py]: remove={} returns a dangerous default value warning.

File: tools/to_ipynb.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    py_filepath = os.path.join(root, path, py_filename)
+    print('Converting {}'.format(py_filepath))
+
+    ipynb_filename = py_filename.split('.')[0] + '.ipynb'
+    ipynb_filepath = os.path.join(root, path, ipynb_filename)
+
+    with open(py_filepath, 'r') as f:
```
Comment (gogasca) [inline tools/to_ipynb.py]: rename f to input_file? or any other more descriptive name?

File: census/estimator/trainer/task.ipynb (hunk 3 above)
Comment (dizcology) [reply census/estimator/trainer/task.ipynb]: If the original .py file has top level comments they will be turned into markdown cells, so the burden is on the .py file to have better documentation and comments.  See for example the census/keras notebook i just added.

File: census/estimator/trainer/task.py (hunk 1 above)
Comment (dizcology) [reply census/estimator/trainer/task.py]: okay - will wait until your PR is merged and then update this.

File: census/estimator/trainer/task.py (hunk 2 above)
Comment (dizcology) [reply census/estimator/trainer/task.py]: same

File: tools/to_ipynb.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+import re
+from redbaron import RedBaron
+import yaml
+
+
+with open('samples.yaml', 'r') as f:
+    samples = yaml.load(f.read())
+
+
+def should_concat(prev_type, cur_type):
```
[the remark this answers is not in the document]
Comment (dizcology) [reply tools/to_ipynb.py]: done

File: tools/to_ipynb.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+    concat_types = ['comment', 'import', 'from_import', 'assignment']
+    import_types = ['import', 'from_import']
+
+    result = False
+
+    if prev_type == cur_type and cur_type in concat_types:
+        result = True
```
[the remark this answers is not in the document]
Comment (dizcology) [reply tools/to_ipynb.py]: done

File: tools/to_ipynb.py [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+
+    if prev_type in import_types and cur_type in import_types:
+        result = True
```
[the remark this answers is not in the document]
Comment (dizcology) [reply tools/to_ipynb.py]: done

File: tools/to_ipynb.py (hunk 5 above)
Comment (dizcology) [reply tools/to_ipynb.py]: changed to `remove=None`, but why is that dangerous?

File: tools/to_ipynb.py (hunk 6 above)
Comment (dizcology) [reply tools/to_ipynb.py]: done

File: tools/to_ipynb.py (hunk 4 above)
Comment (dizcology) [reply tools/to_ipynb.py]: feel free to point out issues.

File: tools/git_clone.p [hunk 10]
```diff
@@ -0,0 +1,4 @@
+import sys
```
[the remark this answers is not in the document]
Comment (dizcology) [reply tools/git_clone.p]: comment added.  This file really is a template to be re-formatted and put into a cell.

File: census/estimator/trainer/task.py (hunk 1 above)
Comment (gogasca) [reply census/estimator/trainer/task.py]: PR merged

File: census/estimator/trainer/task.py (hunk 2 above)
Comment (gogasca) [reply census/estimator/trainer/task.py]: PR merged

**Pull request merged**